File: 2023/puzzle3.hs
import System.IO ()
import Data.Char (isDigit)

main :: IO ()
main = do
    lines <- readLinesFromFile "puzzle3.dat"
    let numbers = concatMap (uncurry parseNumbers) (zip [0..] lines)
        partNumbersTuples = filter (\(row, col, len, num) -> checkSurrounding (row, col, len, num) lines) numbers
        partNumbers = map (\(row, col, len, num) -> num) partNumbersTuples
    putStrLn $ "Part 1: " ++ show (sum partNumbers)

-- function that takes a (row, col, len, num) tuple and a list of lines and calls
-- getSurroundingCoords to get the surrounding coordinates of the number
-- and returns true if any of the surrounding characters are not digits and are not '.'
checkSurrounding :: (Int, Int, Int, Int) -> [String] -> Bool
checkSurrounding (row, col, len, num) lines = do
    let surroundingCoords = getSurroundingCoords (row, col, len, num) (length lines, length (head lines))
        surroundingChars = map (\(r, c) -> lines !! r !! c) surroundingCoords
    any (\x -> not (isDigit x) && x /= '.') surroundingChars

-- function takes a file path and returns a list of lines from the file
readLinesFromFile :: FilePath -> IO [String]
readLinesFromFile filePath = do
    contents <- readFile filePath
    return (lines contents)

-- Function to get surrounding coordinates
getSurroundingCoords :: (Int, Int, Int, Int) -> (Int, Int) -> [(Int, Int)]
getSurroundingCoords (row, col, len, num) (height, width) =
    filter withinBounds allSurroundingCoords
  where
    -- Calculate end column of the number
    endCol = col + len - 1

    -- All possible surrounding coordinates
    allSurroundingCoords = [(r, c) | r <- [row - 1, row, row + 1],
                                     c <- [col - 1, col..endCol + 1],
                                     not (r == row && c >= col && c <= endCol)]

    -- Check if a coordinate is within the grid boundaries
    withinBounds (r, c) = r >= 0 && r < height && c >= 0 && c < width


-- Main function to process the string, now takes an additional row number
parseNumbers :: Int -> String -> [(Int, Int, Int, Int)]
parseNumbers rowNum str = parseHelper str 0
  where
    parseHelper :: String -> Int -> [(Int, Int, Int, Int)]
    parseHelper [] _ = []
    parseHelper (x:xs) idx
      | isDigit x = let (numStr, rest) = span isDigit (x:xs)
                        num = read numStr
                        len = length numStr
                    in (rowNum, idx, len, num) : parseHelper rest (idx + len)
      | otherwise = parseHelper xs (idx + 1)

File: 2023/puzzle3.py
import re
from itertools import chain

def part1(data, width, height, numberList):
    numbers = [n for n in numberList \
            if any([isSymbol(data[srow][scol]) for srow, scol in getAdjacent(height, width, n)])]
    print ("Part 1:", sum(x[3] for x in numbers))

def part2(data, width, height, numberList):
    # make a list of coordinates of every '*' in the data
    stars = [(row, col) for row, line in enumerate(data) for col, ch in enumerate(line) if ch == '*']
    numberDict = {n: getAdjacent(height, width, n) for n in numberList}
    starDict = {star: [n for n, v in numberDict.items() if star in v] for star in stars}
    print ("Part 2:", sum([x[0][3] * x[1][3] for x in [v for v in starDict.values() if len(v) == 2]]))

# extractNumbers: take a string, extract all the numbers from it, returning a list of tuples containing row,
# column, length, and the number itself as a string
def extractNumbers(line, row):
    return [(row, x[1], len(x[0]), int(x[0])) \
            for x in zip(re.findall(r'\d+', line), [x.start() for x in re.finditer(r'\d+', line)])]

# given a width, a height, and a tuple from extractNumbers, return a list of coordinates adjacent to the number
# string
def getAdjacent(height, width, number):
    adjacentCoords = [(number[0] - 1, number[1] -1 + x) for x in range(number[2]+2)] + \
                    [(number[0] + 1, number[1] -1 + x) for x in range(number[2]+2)] + \
                    [(number[0], number[1] - 1), (number[0], number[1] + number[2])]
    return [x for x in adjacentCoords if x[0] >= 0 and x[0] < height and x[1] >= 0 and x[1] < width]

# return True if the character is not a digit and is not a period
def isSymbol(ch): return not ch.isdigit() and ch != '.'

# readData: read the data file into a list of lists stripping newlines
def readData():
    with open("puzzle3.dat") as f:
        return f.read().splitlines()

if __name__ == "__main__":
    data = readData()
    width, height = len(data[0]), len(data)
    numberList = list(chain(*[extractNumbers(line, row) for row, line in enumerate(data)]))
    part1(data, width, height, numberList)
    part2(data, width, height, numberList)
